File: ruby/easy/minimum_depth_of_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def min_depth(root)
    return 0 unless root
    queue = []
    queue.unshift([root, 0])
    while(!queue.empty?)
        current = queue.pop()
        return current[1]+1 if current[0].left.nil? && current[0].right.nil?
        queue.unshift([current[0].left, current[1]+1]) if current[0].left
        queue.unshift([current[0].right, current[1]+1]) if current[0].right
    end
end

File: javascript/easy/minimumDepthOfBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function(root) {
    if(root == null) return 0
    let queue = []
    queue.unshift({ node: root, depth: 0});
    while(queue.length > 0) {
        let current = queue.pop();
        if(current.node.left == null && current.node.right == null) {
            return current.depth + 1;
        }

        if(current.node.left != null) {
            queue.unshift({ node: current.node.left, depth: current.depth + 1 });
        }

        if(current.node.right != null) {
            queue.unshift({ node: current.node.right, depth: current.depth + 1 });
        }
    }

    return 0;
};
